### DISCUSSION TITLE
Rule is not executed on Switch change

### DISCUSSION THREAD

**Discussion by gon0** [arendst/Tasmota#22642] [Support request] [answered]
Hello everyone,

I am stuck at a point, while using Tasmota to build a Bike blinking and Side marker lights system ([Tasmota 14.3.0 (TasmoCompiler-esp32generic)))

I am using these module parameters
![grafik](https://github.com/user-attachments/assets/e245c48e-e345-4d5a-a851-4252cf03cb26)

I intend to use the switches for:
Switch1 is for activation of **left** blinker
Switch2 is for activation of **right** blinker
Switch3 is for permanent activation of both **left and right** blinkers as permanent-on Side marker lights

Relay1 switches left Side marker lights
Relay2 switches right Side marker lights

I thought of following rules to use the functions of the switches:
```
Rule1 ON Switch1#state=1 do Power1 3 endon [this lets the side marker lights blink to use them as blinkers]
Rule2 ON Switch2#state=1 do Power2 3 endon [this lets the side marker lights blink to use them as blinkers]
Rule3 ON Switch3#state=1 do Power2 1 endon [since Power1 is switched on by default using switch1, I just want to permanently switch on Power2 also. So The side marker lights on both sides are permanently switched on.]

BlinkTime: {"BlinkTime":6}
BlinkCount {"BlinkCount":0}

```

This happens when activating and deactivating Switch1:
```
09:27:29.940 RSL: RESULT = {"POWER1":"ON"}
09:27:29.942 RSL: POWER1 = ON
09:27:35.688 RSL: RESULT = {"POWER1":"OFF"}
09:27:35.690 RSL: POWER1 = OFF
```
This happens when activating and deactivating Switch2:
```
09:28:03.710 RSL: RESULT = {"POWER2":"ON"}
09:28:03.712 RSL: POWER2 = ON
09:28:08.028 RSL: RESULT = {"POWER2":"OFF"}
09:28:08.031 RSL: POWER2 = OFF
```
This happens when activating and deactivating Switch3:
```
09:28:21.897 RSL: RESULT = {"POWER1":"ON"}
09:28:21.899 RSL: POWER1 = ON
09:28:27.088 RSL: RESULT = {"POWER1":"OFF"}
09:28:27.090 RSL: POWER1 = OFF
```

It does not matter which Switch I activate, none of the rules is executed. I do not have a clue, why. Do I use a wrong syntax?

**Comment by barbudor** [comment]
What do you mean "none of the rule are executed" ?
As you show logs that turn power on and turn it off 6 seconds after
Doesn't your lights actually blink based on your blink settings ?

Use `SetLog 3` for more detaillled logs

And check your `BlinkCount`  and `BlinkTime` settings

Note 1 : When you using blinking, Tasmota only reports ON and OFF at the beginning and end of the sequence, not at each blinking

Note 2 : a `Rule` is actually a rule set that can hold multiple "rules" `ON trigger DO something ENDON` so you can put all your rules into ruleset 1

**Reply by barbudor** [reply]
Actually yourlogs seems coherent with `BlinkTime 6`  and `BlinkCount 10`, not `BlickCount 0`

**Reply by sfromis** [reply, marked as answer]
I saw no indication of you making sure to actually have your 3 rule sets enabled in addition to just being stored. Another thing is that having a `Switch1#state==1` rule will only take care of state value 1, but state value 0 will still have the default action, unless you have taken other steps to decouple.

I'd suggest using test rules like `on Switch1#state do var1 %value% endon` to verify that you even get the switch state values you expect, at the right time.

**Reply by gon0** [reply]
Thank you for your quick replys. I apprechiate your help!
@sfromis, you were right, I did not activate the rules 🙈

@barbudor thank you for the hint. Storing all the rules in a ruleset is fine.

I experimented with different rulesets. The aim is, that after stopping the blinker (Power1 or Power2), the side marker light should go back into the state which is defined by Switch3#state.
Both rule ideas I had do not work: The Blinking continues at Power1 when Switch1#state=0 or at Power2 when Switch2#state=0. The blinking continues regardless of the previous Switch3#state. When Switch3#state changes, the Side Marker Lights follow the state of Switch3#state correctly.
```
Idea #1:
Rule3
	ON Switch3#state=1 do backlog Power1 1; Power2 1 endon
	ON Switch3#state=0 do backlog Power1 0; Power2 0 endon
	ON Switch1#state=0 do if Switch3#state=0 Power1 0 else Power1 1 endif endon ON Switch1#state=1 do Power1 3 endon
	ON Switch2#state=0 do if Switch3#state=0 Power2 0 else Power2 1 endif endon ON Switch2#state=1 do Power2 3 endon

Idea #2:
Rule3 
	ON Switch3#state=1 do backlog Power1 1; Power2 1; var3 %value% endon
	ON Switch3#state=0 do backlog Power1 0; Power2 0; var3 %value%  endon
	ON Switch1#state=0 do Power1 var3 endon ON Switch1#state=1 do Power1 3 endon
	ON Switch2#state=0 do Power2 var3 endon ON Switch2#state=1 do Power2 3 endon

```
@sfromis your hint with the "var-checker-function is good, var3 changes accordingly to Switch3#state:

```
21:00:29.486 RSL: RESULT = {"Var3":"0"}
21:00:32.389 RUL: SWITCH3#STATE=1 performs 'backlog Power1 1; Power2 1; var3 1'
21:00:32.401 RSL: RESULT = {"POWER1":"Blink OFF"}
21:00:32.415 RSL: RESULT = {"POWER1":"ON"}
21:00:32.417 RSL: POWER1 = ON
21:00:32.633 RSL: RESULT = {"POWER2":"Blink OFF"}
21:00:32.646 RSL: RESULT = {"POWER2":"ON"}
21:00:32.648 RSL: POWER2 = ON
21:00:32.883 RSL: RESULT = {"Var3":"1"}
21:00:34.267 CMD: var3
21:00:34.271 RSL: RESULT = {"Var3":"1"}
21:00:37.776 RUL: SWITCH3#STATE=0 performs 'backlog Power1 0; Power2 0; var3 0'
21:00:37.788 RSL: RESULT = {"POWER1":"OFF"}
21:00:37.790 RSL: POWER1 = OFF
21:00:38.016 RSL: RESULT = {"POWER2":"OFF"}
21:00:38.017 RSL: POWER2 = OFF
21:00:38.264 RSL: RESULT = {"Var3":"0"}
```

I am not sure, how to revert the previous state of the respective side marker light when the blinker function is turned off.

**Reply by sfromis** [reply]
> `ON Switch1#state=0 do if Switch3#state=0 ...`

No. You can only use the trigger expressions with `#` in the trigger part of rules, before `do`. Inside of the rule (after `do`), you can use `%switch1%` etc. And you must have () around your conditions with `if`

**Reply by gon0** [reply]
Excellent help, thank you very much, @sfromis !

This works fine:

```
Rule3
	ON Switch3#state=1 do backlog Power1 1; Power2 1 endon
	ON Switch3#state=0 do backlog Power1 0; Power2 0 endon
	ON Switch1#state=0 do if (%switch3%==0) Power1 0 else Power1 1 endif endon ON Switch1#state=1 do Power1 3 endon
	ON Switch2#state=0 do if (%switch3%==0) Power2 0 else Power2 1 endif endon ON Switch2#state=1 do Power2 3 endon
```

Have a great day!
